### ISSUE TITLE
Named feature support for VW buggy

### ISSUE THREAD

**Issue by jonastim** [VowpalWabbit/coba#22]
Hi!
When passing context features as a dictionary with the keys being their names they don't seem to be processed properly. 
While stepping through the code I believe the issue is in `_prep_namespaces` logic
```
for k,v in compress(zip(K,V),V):
    if v.__class__ is str:
        d[f"{k}={v}"] = 1
    else:
        d[k] = v
```

As an example I created this simple custom environment in which the reward is influenced by one of three features:
```
class CustomEnvironment(LambdaSimulation):
    def __init__(self, n_interactions: Optional[int] = 1000):
        super().__init__(n_interactions, self.context, self.actions, self.rewards)
        self.r = CobaRandom(1)


    def actions(self, index: int, context: Context) -> Sequence[Action]:
        """
        actions: A function that should return all valid actions for a given index, context and random state.
        """
        return [0,1]


    def context(self, index: int) -> Context:
        # return tuple([self.r.randoms(1)[0], self.r.randoms(1)[0], self.r.randoms(1)[0]])
        return {
            "feature_1": self.r.randoms(1)[0],
            "feature_2": self.r.randoms(1)[0],
            "feature_3": self.r.randoms(1)[0]
        }

    def rewards(self, index: int, context: Context, action: Action) -> float:
        # reward_probabilities_for_actions = [
        #     0.5 + 1.0 * (context[0] - 0.5),
        #     0.5
        # ]
        reward_probabilities_for_actions = [
            0.5 + 1.0 * (context["feature_1"] - 0.5),
            0.5
        ]
        return np.random.binomial(1, reward_probabilities_for_actions[action])

environments = Environments([CustomEnvironment(5000)]).shuffle([1,2,3,4,5,6,7,8])
learners     = [
    VowpalBagLearner(features=[1, 'x', 'a', 'ax']),
    VowpalSoftmaxLearner(features=[1, 'x', 'a', 'ax']),
    LinUCBLearner(features=[1, 'x', 'a', 'ax']),
    RandomLearner(),
]
result = Experiment(environments, learners).run()
```

When passing the features as a tuple (commented out code) the results look reasonable (albeit not great) 

![tuple_feature](https://user-images.githubusercontent.com/1749498/211953914-29e9f9b2-2d39-4a92-a94a-95ee1e40e971.jpg)

but when passing the features as a dict the performance is random.

![named_feature](https://user-images.githubusercontent.com/1749498/211953908-1963590a-9e9c-403b-b864-36c9ab146e9c.jpg)

The issue seems to be that the value of each numerical feature is set to `1`

<img width="1403" alt="Screenshot 2023-01-11 at 1 43 46 PM" src="https://user-images.githubusercontent.com/1749498/211953300-18d9f151-fbbb-4268-833e-51b9a6a42b87.png">

because of this logic (intended for categorical features?)
<img width="1632" alt="Screenshot 2023-01-11 at 1 42 16 PM" src="https://user-images.githubusercontent.com/1749498/211953515-1d69eb3d-ab77-4b31-8ac0-4bedd1aca64e.png">

Can you confirm that this is a bug and not a misunderstanding on my side?

Thanks,
Jonas

**Comment by mrucker** [comment]
You're absolutely correct. Nice sleuthing and debugging. What I can't seem to figure out is why this is happening for you.

I ran your code with on my machine and I didn't have any problems:

![image](https://user-images.githubusercontent.com/966578/211973153-1e6efe59-aeb5-4ea6-a832-ff1158fa95f3.png)

Unfortunately, the smoking gun is just out of the picture on your second screen shot.

The feature processing should have fallen into an if statement on line 145/146 instead of the else block that you are inside of.

Do you know what version of coba you are on? You can check either by looking in your python environment or running:

```
from coba import __version__

print(__version__)
```

**Comment by jonastim** [comment]
Thanks for the quick response, Mark!
I worked directly in the checked out source code from early November that shows version 5.1.0.
With the latest master the issue is resolved 🙌

The previous instance check seemed to have failed for `feats` as `HashableDict`
 
<img width="932" alt="Screenshot 2023-01-12 at 8 24 32 AM" src="https://user-images.githubusercontent.com/1749498/212148476-9f57e795-7d8b-46f7-8566-920d17b51dfa.png">

The LinUCB implementation seems to be struggling now, though. 
If the features are passed as a tuple LinUCB doesn't even show up in the result plot.
<img width="750" alt="tuple_reward_function" src="https://user-images.githubusercontent.com/1749498/212149056-59246058-e26f-4755-a4ec-d4f856e45aba.png">

and with the named features it performs on par with random

<img width="731" alt="named_reward_function" src="https://user-images.githubusercontent.com/1749498/212149146-68c1c001-b8ae-45ed-a88f-987157b983b2.png">

With simplified, static reward probabilities
```
reward_probabilities_for_actions = [
    0.2,
    0.1
]
```
it actually picked the lower performing arm (even though features are irrelevant in this case)
<img width="737" alt="named_static_reward" src="https://user-images.githubusercontent.com/1749498/212149585-e3ccc0f9-df2f-4a9b-811d-d9d33f3e8016.png">

I am mostly interested in VW and only use LinUCB as the baseline but wanted to raise this nevertheless. 

As a more general question, I was wondering what your experience with VW has been on these types of synthetic data tests. 
I am a bit concerned for example that the softmax explorer wasn't that much better than random in picking a variant that's twice as good. 
The bag explorer performed on par with random when the reward heavily relied on a feature `0.5 + 1.0 * (context["feature_1"] - 0.5)` which you'd imagine would be easy to pick up (much easier than real world examples in which a variant is maybe 1-5% better).

Do you have any guidance on algo, hyperparameter and interaction terms selection that you've seen perform consistently well in these scenarios?

Thanks a lot!

**Comment by mrucker** [comment]
Great catch. You were correct there was a bug in LinUCB.

I recently upgraded coba to support continuous actions and during that refactor I broke LinUCB.

I added a unittest so hopefully it never happens again and released the patch. If you upgrade to 6.2.2 LinUCB is now working.

**Comment by jonastim** [comment]
Just gave the latest code a try and the performance looks much better for tuple-based features 
<img width="733" alt="tuple" src="https://user-images.githubusercontent.com/1749498/212203898-54f4c416-7240-48e9-a31e-a4fc9d6d1ee0.png">
but the named feature support still seems broken
<img width="735" alt="named_feature" src="https://user-images.githubusercontent.com/1749498/212203940-fadd59b5-42e6-40fa-90be-1d531e59bee9.png">


Code for reference
```

from typing import Optional, Sequence

import numpy as np

from coba import CobaRandom, Environments, RandomLearner, Experiment, VowpalSoftmaxLearner, VowpalBagLearner
from coba.environments import LambdaSimulation
from coba.primitives import Context, Action


class CustomEnvironment(LambdaSimulation):
    def __init__(self, n_interactions: Optional[int] = 1000):
        super().__init__(n_interactions, self.context, self.actions, self.rewards)
        self.r = CobaRandom(1)

    def actions(self, index: int, context: Context) -> Sequence[Action]:
        """
        actions: A function that should return all valid actions for a given index, context and random state.
        """
        return [0, 1]

    def context(self, index: int) -> Context:
        # return tuple([self.r.randoms(1)[0], self.r.randoms(1)[0], self.r.randoms(1)[0]])
        return {
            "feature_1": self.r.randoms(1)[0],
            "feature_2": self.r.randoms(1)[0],
            "feature_3": self.r.randoms(1)[0]
        }

    def rewards(self, index: int, context: Context, action: Action) -> float:
        reward_probabilities_for_actions = [
            0.2,
            0.1
        ]
        # reward_probabilities_for_actions = [
        #     0.5 + 1.0 * (context["feature_1"] - 0.5),
        #     0.5
        # ]
        return np.random.binomial(1, reward_probabilities_for_actions[action])


from coba import LinUCBLearner

environments = Environments([CustomEnvironment(5000)]).shuffle([1, 2, 3, 4, 5, 6, 7, 8])
learners = [
    VowpalBagLearner(features=[1, 'x', 'a', 'ax']),
    VowpalSoftmaxLearner(features=[1, 'x', 'a', 'ax']),
    LinUCBLearner(features=[1, 'x', 'a', 'ax']),
    RandomLearner(),
]
result = Experiment(environments, learners).run()

result.plot_learners()
```

**Comment by mrucker** [comment]
Unfortunately LinUCB doesn't currently support named features. It's easy to miss it but in the output log it should say:

![image](https://user-images.githubusercontent.com/966578/212208243-d99fbbf7-52a7-4a75-ba11-498d58228d07.png)

**Comment by mrucker** [comment]
(Also, regarding your question about VW and hyperparameters, yeah it is hard to know how to tune them. Unfortunately VW doesn't have the best documentation. I can only speak to my personal experience, but VW really shines when I'm either doing off policy learning or working with very messy real world data. For example data with covariate or concept shifts. It is fairly easy to emulate that kind of data in coba but if all you really want is to test linear IID datasets I feel like it will probably be difficult to beat UCB methods.)

**Comment by mrucker** [comment]
(Also Also, LinUCB is a good baseline. Another good baseline is simply the UcbBanditLearner. So long as you are working with actions that don't have features that should give you the performance of the best constant predictor. That is, it learns the best action and only plays that regardless of context.)

**Comment by jonastim** [comment]
Ah, I missed that log line and I guess was a bit surprised as it was working before.

Thanks for your help! I'll keep exploring the different algos and hyperparameters. If you have any guidance on what worked well for you it'd be much appreciated.

**Comment by mrucker** [comment]
Yeah, its easy to miss. I even had a hard time finding it and I was looking for it. I could maybe make that more readable too.

If you end up needing/wanting LinUCB to work with dicts of features it wouldn't be super hard to get that functionality in place. We'd just need to implement some kind of hashing trick. I'd probably do it as an environment filter so that every learner could take advantage of it and not just LinUCB.

Awesome! Yes please reach out if you have more questions. And yeah, sorry to hear VW has been kind of touchy. I actually don't work for Microsoft so I'm hesitant to say anything about VW and how to tune it though I can say it has been the best learner for me on several of my own projects. I'm actually just a grad student who has created Coba for my own PhD research. So, it could be there is another bug with VW stuff though I've gone over it with a super fine toothed comb and research teams inside of Microsoft use it for their own VW work.

In terms of messier real world data sets I'm sure you also know then that coba can also download any of the data sets from openml.
